src/app/clusters/dishwasher-alarm-server/dishwasher-alarm-server.cpp

```diff
@@ -0,0 +1,458 @@
+    {
+        ChipLogProgress(Zcl, "Dishwasher Alarm: ERR: Latch is not supported");
+        return EMBER_ZCL_STATUS_FAILURE;
+    }
+
+    EmberAfStatus status = Attributes::Latch::Set(endpoint, latch);
+    if (status != EMBER_ZCL_STATUS_SUCCESS)
+    {
+        ChipLogProgress(Zcl, "Dishwasher Alarm: ERR: writing  latch, err:0x%x", status);
+        return status;
+    }
+
+    return status;
+}
+
+EmberAfStatus DishwasherAlarmServer::SetStateValue(EndpointId endpoint, const BitMask<AlarmMap> newState)
+{
+    BitMask<AlarmMap> supported;
+    BitMask<AlarmMap> finalNewState;
+    finalNewState.Set(newState);
+
+    if (GetSupportedValue(endpoint, &supported) || !supported.HasAll(finalNewState))
+    {
+        ChipLogProgress(Zcl, "Dishwasher Alarm: ERR: State is not supported");
+        return EMBER_ZCL_STATUS_FAILURE;
+    }
+
+    BitMask<AlarmMap> mask;
+    if (GetMaskValue(endpoint, &mask) || !mask.HasAll(finalNewState))
+    {
+        ChipLogProgress(Zcl, "Dishwasher Alarm: ERR: State is not supported");
+        return EMBER_ZCL_STATUS_FAILURE;
+    }
+
+    EmberAfStatus status = EMBER_ZCL_STATUS_SUCCESS;
+    BitMask<AlarmMap> currentState;
+    status = Attributes::State::Get(endpoint, &currentState);
+    if (status != EMBER_ZCL_STATUS_SUCCESS)
+    {
+        ChipLogProgress(Zcl, "Dishwasher Alarm: ERR: reading  state, err:0x%x", status);
+        return status;
+    }
+
+    BitMask<AlarmMap> latch;
+    if (GetLatchValue(endpoint, &latch) == EMBER_ZCL_STATUS_SUCCESS)
+    {
+        BitMask<AlarmMap> finalState = SetStateByLatch(latch, currentState, finalNewState);
+        finalNewState.Set(finalState);
+    }
+
+    status = Attributes::State::Set(endpoint, finalNewState);
+    if (status != EMBER_ZCL_STATUS_SUCCESS)
+    {
+        ChipLogProgress(Zcl, "Dishwasher Alarm: ERR: writing  state, err:0x%x", status);
+        return status;
+    }
+
+    // Generate Notify event.
+    BitMask<AlarmMap> becameActive;
+    becameActive.Set(finalNewState).Clear(currentState);
+    BitMask<AlarmMap> becameInactive;
+    becameInactive.Set(currentState).Clear(finalNewState);
+
+    SendNotifyEvent(endpoint, becameActive, becameInactive, finalNewState, mask);
+    return status;
+}
+
+EmberAfStatus DishwasherAlarmServer::ResetStateValue(EndpointId endpoint, const BitMask<AlarmMap> alarms)
+{
+    BitMask<AlarmMap> supported;
+    if (GetSupportedValue(endpoint, &supported) || !supported.HasAll(alarms))
+    {
+        ChipLogProgress(Zcl, "Dishwasher Alarm: ERR: State is not supported");
+        return EMBER_ZCL_STATUS_FAILURE;
+    }
+
+    BitMask<AlarmMap> currentState;
+    if (GetStateValue(endpoint, &currentState) != EMBER_ZCL_STATUS_SUCCESS)
+    {
+        return EMBER_ZCL_STATUS_FAILURE;
+    }
+
+    BitMask<AlarmMap> newState;
+    newState.Set(currentState).Clear(alarms);
+
+    if (Attributes::State::Set(endpoint, newState) != EMBER_ZCL_STATUS_SUCCESS)
```

Comment:
I have to ask:  Why didn't we just:
```
    BitMask<AlarmMap> supported;
    if (GetSupportedValue(endpoint, &supported) || !supported.HasAll(alarms))
    {
        ChipLogProgress(Zcl, "Dishwasher Alarm: ERR: State is not supported");
        return EMBER_ZCL_STATUS_FAILURE;
    }

    BitMask<AlarmMap> state;
    if (GetStateValue(endpoint, &state) != EMBER_ZCL_STATUS_SUCCESS)
    {
        return EMBER_ZCL_STATUS_FAILURE;
    }

    state.Clear(alarms);
    return SetStateValue(state);
```
instead of re-implementing pieces of SetStateValue.

Reply:
SetStateValue has judged whether the Latch attribute is set , so I try to optimize it with your suggestion.